```

"C:\buildbot\slave\windows-msvc-dev\build\target\debug\build\azure-3f01c03fa9e237b5\out\build\install.vcxproj" (default target) (1) ->
"C:\buildbot\slave\windows-msvc-dev\build\target\debug\build\azure-3f01c03fa9e237b5\out\build\ALL_BUILD.vcxproj" (default target) (3) ->
"C:\buildbot\slave\windows-msvc-dev\build\target\debug\build\azure-3f01c03fa9e237b5\out\build\azure.vcxproj" (default target) (4) ->
(ClCompile target) -> 
  C:\Users\Administrator\.cargo\git\checkouts\rust-azure-984e17605d95e219\a2351f2\libazure\DrawTargetD2D.cpp(1294): error C2259: 'mozilla::gfx::PathBuilderD2D': cannot instantiate abstract class [C:\buildbot\slave\windows-msvc-dev\build\target\debug\build\azure-3f01c03fa9e237b5\out\build\azure.vcxproj]
  C:\Users\Administrator\.cargo\git\checkouts\rust-azure-984e17605d95e219\a2351f2\libazure\PathD2D.cpp(348): error C2259: 'mozilla::gfx::PathBuilderD2D': cannot instantiate abstract class [C:\buildbot\slave\windows-msvc-dev\build\target\debug\build\azure-3f01c03fa9e237b5\out\build\azure.vcxproj]
```

--- a/components/canvas/canvas_paint_thread.rs
+++ b/components/canvas/canvas_paint_thread.rs
@@ -173,6 +173,9 @@ impl<'a> CanvasPaintThread<'a> {
                             Canvas2dMsg::ArcTo(ref cp1, ref cp2, radius) => {
                                 painter.arc_to(cp1, cp2, radius)
                             }
+                            Canvas2dMsg::Ellipse(ref center, radius_x, radius_y, rotation, start, end, ccw) => {
+                                painter.ellipse(center, radius_x, radius_y, rotation, start, end, ccw)
+                            }
                             Canvas2dMsg::RestoreContext => painter.restore_context_state(),
                             Canvas2dMsg::SaveContext => painter.save_context_state(),
                             Canvas2dMsg::SetFillStyle(style) => painter.set_fill_style(style),
@@ -499,6 +502,17 @@ impl<'a> CanvasPaintThread<'a> {
             self.arc(&Point2D::new(cx, cy), radius,
                      angle_start, angle_end, anticlockwise);
         }
+    }
+
+    fn ellipse(&mut self,
+           center: &Point2D<AzFloat>,
+           radius_x: AzFloat,
+           radius_y: AzFloat,
+           rotation_angle: AzFloat,
+           start_angle: AzFloat,
+           end_angle: AzFloat,
+           ccw: bool) {
+        self.path_builder.ellipse(*center, radius_x, radius_y, rotation_angle, start_angle, end_angle, ccw);
     }
 
     fn set_fill_style(&mut self, style: FillOrStrokeStyle) {

--- a/components/canvas_traits/canvas.rs
+++ b/components/canvas_traits/canvas.rs
@@ -42,6 +42,7 @@ pub enum Canvas2dMsg {
     ClearRect(Rect<f32>),
     Clip,
     ClosePath,
+    Ellipse(Point2D<f32>, f32, f32, f32, f32, f32, bool),
     Fill,
     FillText(String, f64, f64, Option<f64>),
     FillRect(Rect<f32>),

--- a/components/script/dom/canvasrenderingcontext2d.rs
+++ b/components/script/dom/canvasrenderingcontext2d.rs
@@ -957,6 +957,26 @@ impl CanvasRenderingContext2DMethods for CanvasRenderingContext2D {
         Ok(())
     }
 
+    // https://html.spec.whatwg.org/multipage/#dom-context-2d-ellipse
+    fn Ellipse(&self, x: f64, y: f64, rx: f64, ry: f64, rotation: f64, start: f64, end: f64, ccw: bool) -> ErrorResult {
+        if !([x, y, rx, ry, rotation, start, end].iter().all(|x| x.is_finite())) {
+            return Ok(());
+        }
+        if rx < 0.0 || ry < 0.0 {
+            return Err(Error::IndexSize);
+        }
+
+        let msg = CanvasMsg::Canvas2d(Canvas2dMsg::Ellipse(Point2D::new(x as f32, y as f32),
+                                                       rx as f32,
+                                                       ry as f32,
+                                                       rotation as f32,
+                                                       start as f32,
+                                                       end as f32,
+                                                       ccw));
+        self.ipc_renderer.send(msg).unwrap();
+        Ok(())
+    }
+
     // https://html.spec.whatwg.org/multipage/#dom-context-2d-imagesmoothingenabled
     fn ImageSmoothingEnabled(&self) -> bool {
         let state = self.state.borrow();

--- a/components/script/dom/paintrenderingcontext2d.rs
+++ b/components/script/dom/paintrenderingcontext2d.rs
@@ -263,6 +263,11 @@ impl PaintRenderingContext2DMethods for PaintRenderingContext2D {
         self.context.ArcTo(cp1x, cp1y, cp2x, cp2y, r)
     }
 
+    // https://html.spec.whatwg.org/multipage/#dom-context-2d-ellipse
+    fn Ellipse(&self, x: f64, y: f64, rx: f64, ry: f64, rotation: f64, start: f64, end: f64, ccw: bool) -> ErrorResult {
+        self.context.Ellipse(x, y, rx, ry, rotation, start, end, ccw)
+    }
+
     // https://html.spec.whatwg.org/multipage/#dom-context-2d-imagesmoothingenabled
     fn ImageSmoothingEnabled(&self) -> bool {
         self.context.ImageSmoothingEnabled()

--- a/components/script/dom/webidls/CanvasRenderingContext2D.webidl
+++ b/components/script/dom/webidls/CanvasRenderingContext2D.webidl
@@ -264,4 +264,9 @@ interface CanvasPath {
   // [LenientFloat] void ellipse(double x, double y, double radiusX, double radiusY,
   //                             double rotation, double startAngle, double endAngle,
   //                             boolean anticlockwise);
+
+  [Throws]
+  void ellipse(unrestricted double x, unrestricted double y, unrestricted double radius_x,
+               unrestricted double radius_y, unrestricted double rotation, unrestricted double startAngle,
+               unrestricted double endAngle, optional boolean anticlockwise = false);
 };

--- a/tests/wpt/metadata/html/dom/interfaces.html.ini
+++ b/tests/wpt/metadata/html/dom/interfaces.html.ini
@@ -3927,9 +3927,6 @@
   [CanvasRenderingContext2D interface: attribute direction]
     expected: FAIL
 
-  [CanvasRenderingContext2D interface: operation ellipse(unrestricted double,unrestricted double,unrestricted double,unrestricted double,unrestricted double,unrestricted double,unrestricted double,boolean)]
-    expected: FAIL
-
   [CanvasRenderingContext2D interface: document.createElement("canvas").getContext("2d") must inherit property "width" with the proper type (1)]
     expected: FAIL
 
@@ -4048,12 +4045,6 @@
     expected: FAIL
 
   [CanvasRenderingContext2D interface: document.createElement("canvas").getContext("2d") must inherit property "direction" with the proper type (69)]
-    expected: FAIL
-
-  [CanvasRenderingContext2D interface: document.createElement("canvas").getContext("2d") must inherit property "ellipse" with the proper type (79)]
-    expected: FAIL
-
-  [CanvasRenderingContext2D interface: calling ellipse(unrestricted double,unrestricted double,unrestricted double,unrestricted double,unrestricted double,unrestricted double,unrestricted double,boolean) on document.createElement("canvas").getContext("2d") with too few arguments must throw TypeError]
     expected: FAIL
 
   [CanvasPattern interface: operation setTransform(SVGMatrix)]
@@ -9589,9 +9580,6 @@
     expected: FAIL
 
   [CanvasRenderingContext2D interface: document.createElement("canvas").getContext("2d") must inherit property "direction" with the proper type (68)]
-    expected: FAIL
-
-  [CanvasRenderingContext2D interface: document.createElement("canvas").getContext("2d") must inherit property "ellipse" with the proper type (78)]
     expected: FAIL
 
   [CanvasPattern interface: operation setTransform(DOMMatrixInit)]

--- a/tests/wpt/mozilla/meta/mozilla/css-paint-api/background-image-tiled.html.ini
+++ b/tests/wpt/mozilla/meta/mozilla/css-paint-api/background-image-tiled.html.ini
@@ -1,0 +1,3 @@
+[background-image-tiled.html]
+  type: reftest
+  expected: FAIL

